prov/verbs/src/verbs_cq.c

```diff
@@ -182,97 +202,122 @@ static uint64_t fi_ibv_comp_flags(struct ibv_wc *wc)
 	return flags;
 }
 
-static ssize_t fi_ibv_cq_read_context(struct fid_cq *cq, void *buf, size_t count)
+static void fi_ibv_cq_read_context_entry(struct ibv_wc *wc, int i, void *buf)
 {
-	struct fi_ibv_cq *_cq;
 	struct fi_cq_entry *entry = buf;
-	ssize_t ret = 0, i;
 
-	_cq = container_of(cq, struct fi_ibv_cq, cq_fid);
-	if (_cq->wc.status)
-		return -FI_EAVAIL;
+	entry[i].op_context = (void *) (uintptr_t) wc->wr_id;
+}
 
-	for (i = 0; i < count; i++) {
-		ret = ibv_poll_cq(_cq->cq, 1, &_cq->wc);
-		if (ret <= 0)
-			break;
+static void fi_ibv_cq_read_msg_entry(struct ibv_wc *wc, int i, void *buf)
+{
+	struct fi_cq_msg_entry *entry = buf;
 
-		if (_cq->wc.status) {
-			ret = -FI_EAVAIL;
-			break;
-		}
+	entry[i].op_context = (void *) (uintptr_t) wc->wr_id;
+	entry[i].flags = fi_ibv_comp_flags(wc);
+	entry[i].len = (uint64_t) wc->byte_len;
+}
 
-		entry->op_context = (void *) (uintptr_t) _cq->wc.wr_id;
-		entry += 1;
-	}
+static void fi_ibv_cq_read_data_entry(struct ibv_wc *wc, int i, void *buf)
+{
+	struct fi_cq_data_entry *entry = buf;
 
-	return i ? i : (ret ? ret : -FI_EAGAIN);
+	entry[i].op_context = (void *) (uintptr_t) wc->wr_id;
+	entry[i].flags = fi_ibv_comp_flags(wc);
+
+	entry[i].data = (wc->wc_flags & IBV_WC_WITH_IMM) ?
+		ntohl(wc->imm_data) : 0;
+
+	entry->len = (wc->opcode & (IBV_WC_RECV | IBV_WC_RECV_RDMA_WITH_IMM)) ?
+		wc->byte_len : 0;
 }
 
-static ssize_t fi_ibv_cq_read_msg(struct fid_cq *cq, void *buf, size_t count)
+static int fi_ibv_match_ep_id(struct slist_entry *entry, const void *ep_id)
 {
-	struct fi_ibv_cq *_cq;
-	struct fi_cq_msg_entry *entry = buf;
-	ssize_t ret = 0, i;
+	struct fi_ibv_msg_epe *epe = container_of(entry, struct fi_ibv_msg_epe, entry);
 
-	_cq = container_of(cq, struct fi_ibv_cq, cq_fid);
-	if (_cq->wc.status)
-		return -FI_EAVAIL;
+	if (epe->ep->ep_id == (uint64_t) ep_id)
+		return 1;
 
-	for (i = 0; i < count; i++) {
-		ret = ibv_poll_cq(_cq->cq, 1, &_cq->wc);
-		if (ret <= 0)
-			break;
+	return 0;
+}
 
-		if (_cq->wc.status) {
-			ret = -FI_EAVAIL;
-			break;
-		}
+/* Must call with cq->lock held */
+ssize_t fi_ibv_poll_cq(struct fi_ibv_cq *cq, struct ibv_wc *wc)
+{
+	struct fi_ibv_msg_epe *epe;
+	struct slist_entry *entry;
+	ssize_t ret;
+
+	ret = ibv_poll_cq(cq->cq, 1, wc);
+	if (ret <= 0)
+		return ret;
+
+	if (wc->opcode == IBV_WC_RECV || wc->opcode == IBV_WC_RECV_RDMA_WITH_IMM)
+		return ret;
 
-		entry->op_context = (void *) (uintptr_t) _cq->wc.wr_id;
-		entry->flags = fi_ibv_comp_flags(&_cq->wc);
-		entry->len = (uint64_t) _cq->wc.byte_len;
-		entry += 1;
+	/* TODO Handle the case when app posts a send with same wr_id */
```

Comment:
For now, can you add an assert that the user's wr_id doesn't fall into the masked range?